### DISCUSSION TITLE
Isolating message sub-string for kv-parser based on regex

### DISCUSSION THREAD

**Discussion by amdrsantos** [syslog-ng/syslog-ng#4314] [Q&A] [answered]
I have a set of key values in a log message and I want to extract the key-values and insert them in a structure data. Also, we want to remove the key values from the original message.

But we are having issues isolating the set of key values from the message.

We have added delimiters in the message. We are considering using a template and provide the template to kv-parser.

How can we isolate the key-values from the message?

```
rewrite my_rule {
    subst("!|.*|!", "", value("MESSAGE"));
};

parser p_kv {
    kv-parser (prefix(".SDATA.my-parsed-data."));
};

rewrite r_sequence_id {
    set("${RCPTID}" value(".SDATA.meta.sequenceId"));
}; 

log {
    source(s_src);
    parser(p_kv);
    rewrite(r_sequence_id);
    rewrite(my_rule);
    destination(d_localfile);
};
```

We need something like:
```
parser p_kv {
    kv-parser (prefix(".SDATA.my-parsed-data.") template(<template function>));
};
```

**Comment by bazsi** [comment, marked as answer]
On Mon, Feb 6, 2023, 12:44 Alexandre Santos ***@***.***>
wrote:

> I have a set of key values in a log message and I want to extract the
> key-values and insert them in a structure data. Also, we want to remove the
> key values from the original message.
>
> But we are having issues isolating the set of key values from the message.
>
> We have added delimiters in the message. We are considering using a
> template and provide the template to kv-parser.
>
> How can we isolate the key-values from the message?
>
> rewrite my_rule {
>
>     subst("!|.*|!", "", value("MESSAGE"));
>
> };
>
>
>
> parser p_kv {
>
>     kv-parser (prefix(".SDATA.my-parsed-data."));
>
> };
>
>
>
> rewrite r_sequence_id {
>
>     set("${RCPTID}" value(".SDATA.meta.sequenceId"));
>
> };
>
>
>
> log {
>
>     source(s_src);
>
>     parser(p_kv);
>
>     rewrite(r_sequence_id);
>
>     rewrite(my_rule);
>
>     destination(d_localfile);
>
> };
>
>
> We need something like:
>
> parser p_kv {
>
>     kv-parser (prefix(".SDATA.my-parsed-data.") template(<template function>));
>
> };
>
>
This one should work as you spelled it out.  Doesn't it?

**Reply by amdrsantos** [reply]
I could not find any good template function to get only the string between delimiters.

Let's say the message is: "!|from=<me@example.com>, size=807, nrcpt=1|! This is a test"
What is the template function that I can use to pass "from=<me@example.com>, size=807, nrcpt=1" to the kv-parser

**Reply by bazsi** [reply]
You should first extract it using regexp-parser() into a name-value pair and then you can run kv-parser() on it.

**Reply by amdrsantos** [reply]
Thanks, that was it:
```
parser { regexp-parser(patterns('!@<(?<KVPAIRS>.*)>@!') template("$MESSAGE") prefix(".regexp.")); };
parser { kv-parser (prefix("my-parsed-data") template("${.regexp.KVPAIRS}")); };
```

Is there any way to pass a variable to the prefix? 
Something like:
```
parser {
    regexp-parser(patterns('!@<(?<STDATANAME>\w+) (?<KVPAIRS>.*)>@!') template("$MESSAGE") prefix(".regexp."));
};
parser {
    kv-parser (prefix(".SDATA.${.regexp.STDATANAME}") template("${.regexp.KVPAIRS}"));
};
```
This last example is not working as I get in the structured data: [${.regexp.STDATANAME} from="<me@example.com>" size="807" nrcpt="1"]
The SD-ID is not what I get from the regex.

**Reply by bazsi** [reply]
> Thanks, that was it:
> 
> ```
> parser { regexp-parser(patterns('!@<(?<KVPAIRS>.*)>@!') template("$MESSAGE") prefix(".regexp.")); };
> parser { kv-parser (prefix("my-parsed-data") template("${.regexp.KVPAIRS}")); };
> ```


Cool. Please note that by convention name value pairs starting with the dot are reserved for syslog-ng, you can use anything else here.

Also, I tend to use $1, $2 etc for temporary values as those are excluded in the syslog-ng () driver for example.


> 
> Is there any way to pass a variable to the prefix? Something like:
> 
> ```
> parser {
>     regexp-parser(patterns('!@<(?<STDATANAME>\w+) (?<KVPAIRS>.*)>@!') template("$MESSAGE") prefix(".regexp."));
> };
> parser {
>     kv-parser (prefix(".SDATA.${.regexp.STDATANAME}") template("${.regexp.KVPAIRS}"));
> };
> ```
> 
> This last example is not working as I get in the structured data: [${.regexp.STDATANAME} from="[me@example.com](mailto:me@example.com)" size="807" nrcpt="1"] The SD-ID is not what I get from the regex

Unfortunately no, there's no way currently. A workaround could be to generate a format that syslog-ng can parse and which includes the prefix. (e.g. generate a JSON or a key=value and parse that back)

We can almost do this with map-value-pairs() which takes an input message and reconstructs it, while potentially remapping keys. But that also only supports string literals when rewriting keys.

Adding template support to prefix() arguments is not impossible, at least it always has an input message in context. It is work, as prefix() is a pretty common argument for parsers, and all of them should support it if we add it to one. It's not even very difficult to do.

So one solution is to add this to value-pairs the other is to add it to all the parsers and we might want both.

Do you want to attempt.to.do.this?

**Reply by amdrsantos** [reply]
> Do you want to attempt.to.do.this?
No, not at the moment.
